src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Serialization/JsonSerializerWrapper.Reflection.cs

```diff
@@ -653,5 +664,88 @@ public static JsonSerializerOptions ResolveOptionsInstanceWithSmallBuffer(JsonSe
                 return smallBufferCopy;
             }
         }
+
+        private sealed class Utf8BomInsertingStream : Stream
+        {
+            private const int Utf8BomLength = 3;
+            private readonly static byte[] s_utf8Bom = Encoding.UTF8.GetPreamble();
+
+            private readonly Stream _source;
+            private byte[]? _prefixBytes;
+            private int _prefixBytesOffset = 0;
+            private int _prefixBytesCount = 0;
+
+            public Utf8BomInsertingStream(Stream source)
+            {
+                Debug.Assert(source.CanRead);
+                _source = source;
+            }
+
+            public override bool CanRead => _source.CanRead;
+            public override bool CanSeek => false;
+            public override bool CanWrite => false;
+
+            public override int Read(byte[] buffer, int offset, int count)
+            {
+                if (_prefixBytes is null)
+                {
+                    // This is the first read operation; read the first 3 bytes
+                    // from the source to determine if it already includes a BOM.
+                    // Only insert a BOM if it's missing from the source stream.
+
+                    _prefixBytes = new byte[2 * Utf8BomLength];
+                    int bytesRead = ReadExactlyFromSource(_prefixBytes, Utf8BomLength, Utf8BomLength);
```

Comment:
Isn't this reading bytes 3-6 and not 0-2?

Reply:
Yes. The idea is that it reads the source stream to detect if it already contains a BOM. If it doesn't it will write it to bytes 0-2 and copy the entire buffer to the destination. If it does it will only copy bytes 3-5.

Reply:
Per offline discussion - I was looking at the LOC below `_source.Read(buffer, offset + totalRead, count - totalRead);` and thought that the second parameter was the "src" offset and not the "dest" offset.